src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java

```diff
@@ -0,0 +1,251 @@
+    public Object getFinalOutput() {
+        if (preview == null) {
+            return Collections.emptyList();
+        }
+
+        return preview.outputsOfPatternWithCycle().stream().map(RsApi::getObjectFromResourceAmount).collect(Collectors.toList());
+    }
+
+    public TaskStatus getCraftingTask() {
+        return craftingTask;
+    }
+
+    @Override
+    public boolean cancel() {
+        if (isJobDone() || isJobCanceled()) {
+            return false;
+        }
+        autocraftingComponent.cancel(craftingTask.info().id());
+        return true;
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+        // RS2 does re-create a new instance of the TaskStatus when something changes - so we actually need to re fetch that from the system
+        // It's cursed, but the only way currently
+        if (craftingTask != null) {
+            for (TaskStatus status : autocraftingComponent.getStatuses()) {
+                if (status.info().id() == craftingTask.info().id()) {
+                    this.craftingTask = status;
+                    break;
+                }
+            }
+        }
+        // The following is to get the TaskStatus from the issued task
+        if (futureTask == null || !futureTask.isDone() || craftingTask != null) {
+            return;
+        }
+        Optional<TaskId> optionalId;
+
+        try {
+            optionalId = futureTask.get();
+        } catch (InterruptedException | ExecutionException e) {
+            AdvancedPeripherals.debug("Tried to get the task but the task was not done. Should be done", org.apache.logging.log4j.Level.ERROR);
+            fireEvent(true, StatusConstants.UNKNOWN_ERROR);
+            return;
+        }
+
+        if (optionalId.isEmpty()) {
+            // This indicates that the second calculation was not successful. Well I guess. There is no java doc and I currently
+            // don't get an answer from the maintainer. So maybe we want to fire the crafting event
+            return;
+        }
+
+        TaskId id = optionalId.get();
+        for (TaskStatus status : autocraftingComponent.getStatuses()) {
+            if (status.info().id() == id) {
+                this.craftingTask = status;
+                // And only now we set that the crafting is started.
+                setStartedCrafting();
+                break;
+            }
+        }
+    }
+
+    @Override
+    protected void maybeCraft() {
+        if (startedCrafting || futureTask != null || futureCalculationResult == null || !futureCalculationResult.isDone()) {
+            return;
+        }
+
+        Optional<Preview> optionalPreview;
+
+        try {
+            optionalPreview = futureCalculationResult.get();
+        } catch (ExecutionException | InterruptedException ex) {
+            AdvancedPeripherals.debug("Tried to get preview, but preview calculation is not done. Should be done.", org.apache.logging.log4j.Level.ERROR);
+            ex.printStackTrace();
```

Comment:
I believe log4j's logger can directly log throwable, `printStackTrace` will never saves to log tho.